### PR TITLE
Close #12403: Refactor FILTER_PALETTE_ID to use strong enum

### LINKED ISSUE #12403: Refactor FILTER_PALETTE_ID to use strong enum

### PR DESCRIPTION
**Pull request by Lukasz-Pe**
Refactored usages of FilterPaletteID.

### CODE REVIEW

**Comment by Lukasz-Pe** [comment]
I've changed line 23 in `src/openrct2-ui/drawing/engines/opengl/TextureCache.h` from

`#include <src/openrct2/drawing/Drawing.h>`

to:

`#include "../../../../openrct2/drawing/Drawing.h"`

to test if this compiles.

I do not get why the former has caused compilation errors at d1983cc.

Could someone explain this to me?

**Comment by duncanspumpkin** [comment]
I feel there is too much happening here. We use palette id's and colour id's interchangeably in the low level draw commands selected by a flag in the image id and that leads to code that is very hard to verify and follow. Just from a very quick glance I can see at least one instance where `EnumToFlag` has been used where it should definitely not be used. I'm surprised it hasn't caused a compile error.

**Comment by Lukasz-Pe** [comment]
Could you give me a hint what should I be looking for to correct it?

File: src/openrct2-ui/windows/Ride.cpp [hunk 1]
```diff
@@ -4863,7 +4863,7 @@ static void window_ride_colour_paint(rct_window* w, rct_drawpixelinfo* dpi)
                 int32_t terniaryColour = 0;
                 if (stationObj->Flags & STATION_OBJECT_FLAGS::IS_TRANSPARENT)
                 {
-                    terniaryColour = IMAGE_TYPE_TRANSPARENT | (GlassPaletteIds[trackColour.main] << 19);
+                    terniaryColour = IMAGE_TYPE_TRANSPARENT | (EnumToFlag(GlassPaletteIds[trackColour.main]) << 19);
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/Ride.cpp:4866]: Not a flag

File: src/openrct2-ui/windows/Scenery.cpp [hunk 2]
```diff
@@ -1268,7 +1268,8 @@ void window_scenery_scrollpaint(rct_window* w, rct_drawpixelinfo* dpi, int32_t s
                     }
                     gfx_draw_sprite(&clipdpi, imageId, { 0x2F, spriteTop }, tertiaryColour);
 
-                    imageId = (sceneryEntry->image + 0x40000006) | (GlassPaletteIds[gWindowSceneryPrimaryColour] << 19);
+                    imageId = (sceneryEntry->image + 0x40000006)
+                        | (EnumToFlag(GlassPaletteIds[gWindowSceneryPrimaryColour]) << 19);
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/Scenery.cpp:1272]: not a flag

File: src/openrct2-ui/windows/Scenery.cpp [hunk 3]
```diff
@@ -1328,7 +1329,7 @@ void window_scenery_scrollpaint(rct_window* w, rct_drawpixelinfo* dpi, int32_t s
                 if (scenery_small_entry_has_flag(sceneryEntry, SMALL_SCENERY_FLAG_HAS_GLASS))
                 {
                     imageId = ((sceneryEntry->image + gWindowSceneryRotation) + 0x40000004)
-                        + (GlassPaletteIds[gWindowSceneryPrimaryColour] << 19);
+                        + (EnumToFlag(GlassPaletteIds[gWindowSceneryPrimaryColour]) << 19);
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/Scenery.cpp:1332]: not a flag

File: src/openrct2/paint/tile_element/Paint.LargeScenery.cpp [hunk 4]
```diff
@@ -242,7 +242,7 @@ void large_scenery_paint(paint_session* session, uint8_t direction, uint16_t hei
     {
         if (!track_design_save_contains_tile_element(tileElement))
         {
-            sequenceNum = SPRITE_ID_PALETTE_COLOUR_1(PALETTE_46);
+            sequenceNum = SPRITE_ID_PALETTE_COLOUR_1(EnumToFlag(FilterPaletteID::Palette46));
```
Comment (duncanspumpkin) [inline src/openrct2/paint/tile_element/Paint.LargeScenery.cpp:245]: not a flag

File: src/openrct2/paint/tile_element/Paint.Path.cpp [hunk 5]
```diff
@@ -824,13 +824,13 @@ void path_paint(paint_session* session, uint16_t height, const TileElement* tile
 
         if (!track_design_save_contains_tile_element(tile_element))
         {
-            imageFlags = SPRITE_ID_PALETTE_COLOUR_1(PALETTE_46);
+            imageFlags = SPRITE_ID_PALETTE_COLOUR_1(EnumToFlag(FilterPaletteID::Palette46));
```
Comment (duncanspumpkin) [inline src/openrct2/paint/tile_element/Paint.Path.cpp:827]: not a flag

File: src/openrct2/paint/tile_element/Paint.Path.cpp [hunk 6]
```diff
@@ -824,13 +824,13 @@ void path_paint(paint_session* session, uint16_t height, const TileElement* tile
...
         }
     }
 
     if (session->ViewFlags & VIEWPORT_FLAG_HIGHLIGHT_PATH_ISSUES)
     {
-        imageFlags = SPRITE_ID_PALETTE_COLOUR_1(PALETTE_46);
+        imageFlags = SPRITE_ID_PALETTE_COLOUR_1(EnumToFlag(FilterPaletteID::Palette46));
```
Comment (duncanspumpkin) [inline src/openrct2/paint/tile_element/Paint.Path.cpp:833]: not a flag

File: src/openrct2/paint/tile_element/Paint.SmallScenery.cpp [hunk 7]
```diff
@@ -49,7 +50,7 @@ void scenery_paint(paint_session* session, uint8_t direction, int32_t height, co
     {
         if (!track_design_save_contains_tile_element(tileElement))
         {
-            baseImageid = SPRITE_ID_PALETTE_COLOUR_1(PALETTE_46);
+            baseImageid = SPRITE_ID_PALETTE_COLOUR_1(EnumToFlag(FilterPaletteID::Palette46));
```
Comment (duncanspumpkin) [inline src/openrct2/paint/tile_element/Paint.SmallScenery.cpp:53]: not a flag

File: src/openrct2/paint/tile_element/Paint.Wall.cpp [hunk 8]
```diff
@@ -194,7 +194,7 @@ void fence_paint(paint_session* session, uint8_t direction, int32_t height, cons
     {
         if (!track_design_save_contains_tile_element(tile_element))
         {
-            dword_141F710 = SPRITE_ID_PALETTE_COLOUR_1(PALETTE_46);
+            dword_141F710 = SPRITE_ID_PALETTE_COLOUR_1(EnumToFlag(FilterPaletteID::Palette46));
```
Comment (duncanspumpkin) [inline src/openrct2/paint/tile_element/Paint.Wall.cpp:197]: not a flag

File: src/openrct2/paint/tile_element/Paint.Path.cpp (hunk 6 above)
Comment (Lukasz-Pe) [reply src/openrct2/paint/tile_element/Paint.Path.cpp]: BTW, here the `imageFlags` name is a bit misleading.

**Comment by Lukasz-Pe** [comment]
@duncanspumpkin Thanks!
I've reverted the original include directive.

**Review by tupaschoal — COMMENTED** [review]
There are some errors regarding include in the CI that need to be sorted.

**Comment by Lukasz-Pe** [comment]
> There are some errors regarding include in the CI that need to be sorted.

It is possible to mitigate this as in [this](https://github.com/OpenRCT2/OpenRCT2/pull/13273#issuecomment-715576295) comment.
I didn't want to make this change without prior discussion (and also because till now everything has worked). I also do not understand why after making the changes it stopped to work.

File: src/openrct2-ui/drawing/engines/opengl/OpenGLDrawingEngine.cpp [hunk 9]
```diff
@@ -708,24 +708,24 @@ void OpenGLDrawingContext::DrawSprite(uint32_t image, int32_t x, int32_t y, uint
     bool special = false;
     if (image & IMAGE_TYPE_REMAP_2_PLUS)
     {
-        palettes.x = TextureCache::PaletteToY((image >> 19) & 0x1F);
-        palettes.y = TextureCache::PaletteToY((image >> 24) & 0x1F);
+        palettes.x = TextureCache::PaletteToY(static_cast<FilterPaletteID>((image >> 19) & 0x1F));
```
Comment (duncanspumpkin) [inline src/openrct2-ui/drawing/engines/opengl/OpenGLDrawingEngine.cpp:711]: If it is anded with 0x1F it is a Colour ID. PaletteID's are 0xFF. The problem though is that all colours are Palettes but not all Palettes are colours. I'm still not sure on this at all what the best approach is.

Comment (Lukasz-Pe) [reply src/openrct2-ui/drawing/engines/opengl/OpenGLDrawingEngine.cpp:711]: PalleteToY returns GLint. According to my IDE this was used only with FilterPaletteID enum and thus the function signature was updated to take this enum. There is an option to revert it to earlier state and always cast strong enums to functions type.

**Comment by Lukasz-Pe** [comment]
I need help with this PR. I am unable to understand why there are problems with including `src/openrct2/drawing/Drawing.h` this way ` #include <src/openrct2/drawing/Drawing.h>` and no problems if I include the same file as follows `#include "../../../../openrct2/drawing/Drawing.h"`

And this issue is valid only in `src/openrct2-ui/drawing/engines/opengl/TextureCache.h`

**Comment by tupaschoal** [comment]
Hello @Lukasz-Pe, there are some merge conflicts, can you please [rebase your PR](https://github.com/OpenRCT2/OpenRCT2/wiki/Rebase-and-Sync-fork-with-OpenRCT2)?

**Comment by Lukasz-Pe** [comment]
@tupaschoal done! :)

File: src/openrct2/object/StationObject.cpp [hunk 10]
```diff
@@ -50,14 +50,14 @@ void StationObject::DrawPreview(rct_drawpixelinfo* dpi, int32_t width, int32_t h
...
-    auto tcolour0 = GlassPaletteIds[colour0];
+    auto tcolour0 = static_cast<int>(GlassPaletteIds[colour0]);
 
     uint32_t imageId = BaseImageId;
     uint32_t tImageId = BaseImageId + 16;
     if (Flags & STATION_OBJECT_FLAGS::HAS_PRIMARY_COLOUR)
     {
         imageId |= (colour0 << 19) | IMAGE_TYPE_REMAP;
-        tImageId |= (GlassPaletteIds[tcolour0] << 19) | IMAGE_TYPE_TRANSPARENT;
+        tImageId |= (static_cast<int>(GlassPaletteIds[tcolour0]) << 19) | IMAGE_TYPE_TRANSPARENT;
```
Comment (tupaschoal) [inline src/openrct2/object/StationObject.cpp:60]:
```suggestion
        tImageId |= (EnumValue(GlassPaletteIds[tcolour0]) << 19) | IMAGE_TYPE_TRANSPARENT;
```

File: src/openrct2/paint/VirtualFloor.cpp [hunk 11]
```diff
@@ -404,7 +405,7 @@ void virtual_floor_paint(paint_session* session)
     if (!weAreOccupied && !weAreLit && weAreAboveGround && weAreOwned)
     {
         int32_t imageColourFlats = SPR_G2_SURFACE_GLASSY_RECOLOURABLE | IMAGE_TYPE_REMAP | IMAGE_TYPE_TRANSPARENT
-            | PALETTE_WATER << 19;
+            | static_cast<int>(FilterPaletteID::PaletteWater) << 19;
```
Comment (tupaschoal) [inline src/openrct2/paint/VirtualFloor.cpp:408]:
```suggestion
            | EnumValue(FilterPaletteID::PaletteWater) << 19;
```

File: src/openrct2/paint/tile_element/Paint.Entrance.cpp [hunk 12]
```diff
@@ -81,7 +81,7 @@ static void ride_entrance_exit_paint(paint_session* session, uint8_t direction,
     uint32_t transparant_image_id = 0, image_id = 0;
     if (stationObj->Flags & STATION_OBJECT_FLAGS::IS_TRANSPARENT)
     {
-        colour_1 = GlassPaletteIds[ride->track_colour[0].main];
+        colour_1 = static_cast<uint8_t>(GlassPaletteIds[ride->track_colour[0].main]);
```
Comment (tupaschoal) [inline src/openrct2/paint/tile_element/Paint.Entrance.cpp:84]:
```suggestion
        colour_1 = EnumValue(GlassPaletteIds[ride->track_colour[0].main]);
```

File: src/openrct2/paint/tile_element/Paint.LargeScenery.cpp [hunk 13]
```diff
@@ -242,7 +242,7 @@ void large_scenery_paint(paint_session* session, uint8_t direction, uint16_t hei
     {
         if (!track_design_save_contains_tile_element(tileElement))
         {
-            sequenceNum = SPRITE_ID_PALETTE_COLOUR_1(PALETTE_46);
+            sequenceNum = SPRITE_ID_PALETTE_COLOUR_1(static_cast<int>(FilterPaletteID::Palette46));
```
Comment (tupaschoal) [inline src/openrct2/paint/tile_element/Paint.LargeScenery.cpp:245]:
```suggestion
            sequenceNum = SPRITE_ID_PALETTE_COLOUR_1(EnumValue(FilterPaletteID::Palette46));
```

File: src/openrct2/paint/tile_element/Paint.Path.cpp [hunk 14]
```diff
@@ -826,13 +826,13 @@ void path_paint(paint_session* session, uint16_t height, const TileElement* tile
 
         if (!track_design_save_contains_tile_element(tile_element))
         {
-            imageFlags = SPRITE_ID_PALETTE_COLOUR_1(PALETTE_46);
+            imageFlags = SPRITE_ID_PALETTE_COLOUR_1(static_cast<int>(FilterPaletteID::Palette46));
```
Comment (tupaschoal) [inline src/openrct2/paint/tile_element/Paint.Path.cpp:829]:
```suggestion
            imageFlags = SPRITE_ID_PALETTE_COLOUR_1(EnumValue(FilterPaletteID::Palette46));
```

File: src/openrct2/paint/tile_element/Paint.Path.cpp [hunk 15]
```diff
@@ -826,13 +826,13 @@ void path_paint(paint_session* session, uint16_t height, const TileElement* tile
...
         }
     }
 
     if (session->ViewFlags & VIEWPORT_FLAG_HIGHLIGHT_PATH_ISSUES)
     {
-        imageFlags = SPRITE_ID_PALETTE_COLOUR_1(PALETTE_46);
+        imageFlags = SPRITE_ID_PALETTE_COLOUR_1(static_cast<int>(FilterPaletteID::Palette46));
```
Comment (tupaschoal) [inline src/openrct2/paint/tile_element/Paint.Path.cpp:835]:
```suggestion
        imageFlags = SPRITE_ID_PALETTE_COLOUR_1(EnumValue(FilterPaletteID::Palette46));
```

File: src/openrct2/paint/tile_element/Paint.SmallScenery.cpp [hunk 16]
```diff
@@ -49,7 +50,7 @@ void scenery_paint(paint_session* session, uint8_t direction, int32_t height, co
     {
         if (!track_design_save_contains_tile_element(tileElement))
         {
-            baseImageid = SPRITE_ID_PALETTE_COLOUR_1(PALETTE_46);
+            baseImageid = SPRITE_ID_PALETTE_COLOUR_1(static_cast<int>(FilterPaletteID::Palette46));
```
Comment (tupaschoal) [inline src/openrct2/paint/tile_element/Paint.SmallScenery.cpp:53]:
```suggestion
            baseImageid = SPRITE_ID_PALETTE_COLOUR_1(EnumValue(FilterPaletteID::Palette46));
```

File: src/openrct2/paint/tile_element/Paint.SmallScenery.cpp [hunk 17]
```diff
@@ -165,8 +166,8 @@ void scenery_paint(paint_session* session, uint8_t direction, int32_t height, co
         {
             // Draw translucent overlay:
             // TODO: Name palette entries
-            int32_t image_id = (baseImageid & 0x7FFFF) + (GlassPaletteIds[sceneryElement->GetPrimaryColour()] << 19)
-                + 0x40000004;
+            int32_t image_id = (baseImageid & 0x7FFFF)
+                + (static_cast<int>(GlassPaletteIds[sceneryElement->GetPrimaryColour()]) << 19) + 0x40000004;
```
Comment (tupaschoal) [inline src/openrct2/paint/tile_element/Paint.SmallScenery.cpp:170]:
```suggestion
                + (EnumValue(GlassPaletteIds[sceneryElement->GetPrimaryColour()]) << 19) + 0x40000004;
```

File: src/openrct2/paint/tile_element/Paint.Surface.cpp [hunk 18]
```diff
@@ -1312,7 +1312,7 @@ void surface_paint(paint_session* session, uint8_t direction, uint16_t height, c
         }
 
         const int32_t image_id = (SPR_WATER_MASK + image_offset) | IMAGE_TYPE_REMAP | IMAGE_TYPE_TRANSPARENT
-            | PALETTE_WATER << 19;
+            | static_cast<int>(FilterPaletteID::PaletteWater) << 19;
```
Comment (tupaschoal) [inline src/openrct2/paint/tile_element/Paint.Surface.cpp:1315]:
```suggestion
            | EnumValue(FilterPaletteID::PaletteWater) << 19;
```

File: src/openrct2/paint/tile_element/Paint.Wall.cpp [hunk 19]
```diff
@@ -194,7 +194,7 @@ void fence_paint(paint_session* session, uint8_t direction, int32_t height, cons
     {
         if (!track_design_save_contains_tile_element(tile_element))
         {
-            dword_141F710 = SPRITE_ID_PALETTE_COLOUR_1(PALETTE_46);
+            dword_141F710 = SPRITE_ID_PALETTE_COLOUR_1(static_cast<int>(FilterPaletteID::Palette46));
```
Comment (tupaschoal) [inline src/openrct2/paint/tile_element/Paint.Wall.cpp:197]:
```suggestion
            dword_141F710 = SPRITE_ID_PALETTE_COLOUR_1(EnumValue(FilterPaletteID::Palette46));
```

File: src/openrct2/ride/gentle/Dodgems.cpp [hunk 20]
```diff
@@ -32,7 +33,8 @@ static void paint_dodgems_roof(paint_session* session, int32_t height, int32_t o
     uint32_t image_id = (SPR_DODGEMS_ROOF_FRAME + offset) | session->TrackColours[SCHEME_TRACK];
     PaintAddImageAsParent(session, image_id, 0, 0, 32, 32, 2, height);
 
-    image_id = (SPR_DODGEMS_ROOF_GLASS + offset) | (PALETTE_DARKEN_3 << 19) | IMAGE_TYPE_TRANSPARENT;
+    image_id = (SPR_DODGEMS_ROOF_GLASS + offset) | (static_cast<int>(FilterPaletteID::PaletteDarken3) << 19)
```
Comment (tupaschoal) [inline src/openrct2/ride/gentle/Dodgems.cpp:36]:
```suggestion
    image_id = (SPR_DODGEMS_ROOF_GLASS + offset) | (EnumValue(FilterPaletteID::PaletteDarken3) << 19)
```

File: src/openrct2/ride/transport/MiniatureRailway.cpp [hunk 21]
```diff
@@ -629,8 +629,9 @@ static void paint_miniature_railway_track_flat(
                 imageId, session->PathElementOnSameHeight, direction, session->CurrentRotation);
             PaintAddImageAsChildRotated(
                 session, direction,
-                (imageIdAlt & 0x7FFFF) | IMAGE_TYPE_REMAP | IMAGE_TYPE_TRANSPARENT | (PALETTE_DARKEN_2 << 19), 0, 6, 32, 20, 2,
-                height, 0, 6, height + 2);
+                (imageIdAlt & 0x7FFFF) | IMAGE_TYPE_REMAP | IMAGE_TYPE_TRANSPARENT
+                    | (static_cast<int>(FilterPaletteID::PaletteDarken2) << 19),
```
Comment (tupaschoal) [inline src/openrct2/ride/transport/MiniatureRailway.cpp:633]:
```suggestion
                    | (EnumValue(FilterPaletteID::PaletteDarken2) << 19),
```

File: src/openrct2/ride/transport/MiniatureRailway.cpp [hunk 22]
```diff
@@ -653,8 +654,9 @@ static void paint_miniature_railway_track_flat(
                 imageId, session->PathElementOnSameHeight, direction, session->CurrentRotation);
             PaintAddImageAsParentRotated(
                 session, direction,
-                (imageIdAlt & 0x7FFFF) | IMAGE_TYPE_REMAP | IMAGE_TYPE_TRANSPARENT | (PALETTE_DARKEN_2 << 19), 0, 6, 32, 20, 2,
-                height, 0, 6, height + 2);
+                (imageIdAlt & 0x7FFFF) | IMAGE_TYPE_REMAP | IMAGE_TYPE_TRANSPARENT
+                    | (static_cast<int>(FilterPaletteID::PaletteDarken2) << 19),
```
Comment (tupaschoal) [inline src/openrct2/ride/transport/MiniatureRailway.cpp:658]:
```suggestion
                    | (EnumValue(FilterPaletteID::PaletteDarken2) << 19),
```

**Comment by tupaschoal** [comment]
Hello @Lukasz-Pe , there are some merge conflicts, can you please rebase and solve them? Thanks!

**Comment by Lukasz-Pe** [comment]
@tupaschoal done!
should I rebase on daily or weekly basis? Or some other schedule shall be used?

**Comment by tupaschoal** [comment]
There's no defined schedule, I ask to rebase before I start reviewing because things might change considerably, depending on the merge conflict. Also because if there are no merge conflicts and I find it good to go, I can merge it right away

**Review by tupaschoal — APPROVED** [review]

**Pull request merged**